### DISCUSSION TITLE
Product Collection Extensibility: Introduce Block-Specific Query Hooks

### DISCUSSION THREAD

**Discussion by xristos3490** [woocommerce/woocommerce#51018] [WooCommerce, Blocks, and Site Editing]
### Background

The _Product Collection_ block's query functions similarly to the _Query Loop_ block. It uses `WP_Query` to display the frontend and sends a REST request to preview products in the Site Editor.

### 1. The Frontend Query

The frontend query is managed via `WP_Query`, utilizing the [build_query_vars_from_query_block()](https://developer.wordpress.org/reference/functions/build_query_vars_from_query_block/) function, and exposes the [query_loop_block_query_vars](https://developer.wordpress.org/reference/hooks/query_loop_block_query_vars/) filter hook.

#### Example:
Here's how third-party developers can currently use the `query_loop_block_query_vars` filter to modify the query for the _Product Collection_ block:

```php
add_filter(
	'query_loop_block_query_vars',
	function ( $query, $block, $page ) {

		$is_product_collection_block = $block->context['query']['isProductCollectionBlock'] ?? false;
		if ( ! $is_product_collection_block ) {
			return $query;
		}

		// Modify query.

		return $query;
	},
	11,
	3
);
```

#### Proposal A: Introduce the `product_collection_block_query_vars` Filter.

The Product Collection block was inspired by the Query Loop block, but it's a separate block. To make things easier for third-party developers and to be more specific, we propose introducing a dedicated filter called `product_collection_block_query_vars` to manage the required query context and reduce unnecessary code.

#### Example:
```php
add_filter(
	'product_collection_block_query_vars',
	function ( $query, $block, $page ) {

		// Modify query.

		return $query;
	},
	10,
	3
);
```

### 2. The Editor's Preview Query

For the Site Editor preview, the block makes a REST request to the `/wp/v2/posts` endpoint. The Product Collection block uses the [rest_product_query](https://developer.wordpress.org/reference/hooks/rest_this-post_type_query/) filter hook to adjust the preview query.

#### Example:
Here’s how third-party developers can currently use the `rest_product_query` filter to customize the preview query for the Product Collection block:

```php
add_filter(
	'rest_product_query',
	function ( $query, $request ) {
		$is_product_collection_block = $request->get_param( 'isProductCollectionBlock' );
		if ( ! $is_product_collection_block ) {
			return $query;
		}

		// Modify query.

		return $query;
	},
	11,
	2
);
```

#### Proposal B: Introduce the `rest_product_collection_block_query_args` Filter.

To further streamline REST query adjustments for third-party developers, we suggest introducing a new filter, `rest_product_collection_block_query_args`, specifically for the Product Collection block.

#### Example:
```php
add_filter(
	'rest_product_collection_block_query_args',
	function ( $query, $request ) {

		// Modify query.

		return $query;
	},
	10,
	2
);
```

These new hooks should improve the query handling of the _Product Collection_ block, reducing the need for boilerplate code.

cc @imanish003 @kmanijak @ObliviousHarmony

**Comment by imanish003** [comment]
Hey @xristos3490 ! Thanks for starting this discussion. It's timely and will be great for DX 💜 
I am sharing some thoughts. But I just wanted to say that in 1st iteration, we can keep it as simple as possible and then iterate 🙇‍♂️

> Proposal A: Introduce the product_collection_block_query_vars Filter.
> 

This will be great as 3PDs won't need to check for `isProductCollectionBlock` context.

- How do you feel about also passing location?
- I am not sure if we need to pass $page as 3rd param. I believe we should set `paged` to `$page` in $query itself. We are doing it [here](https://github.com/woocommerce/woocommerce/blob/4253f2c10a985cdde05159eb27fbbe464a01f7d4/plugins/woocommerce/src/Blocks/BlockTypes/ProductCollection.php#L638) in Product Collection block. Does it make sense?
- The third argument could be more generic and may include location as well. In the future, I believe we should be able to insert collection-specific data into the third argument. However, I am curious about what else we can pass in this third argument.
    - For example, any collection-specific attributes?
    - I don't mind shipping the first iteration without the third argument, considering that we agree to remove the `$page` as the third argument.

Apart from above, I am wondering what could be other names for this hook?

- Should we consider adding `wc_` prefix to avoid any conflicts with other plugins/themes? i.e. `wc_product_collection_block_query_vars`
- Should we consider adding `wp_query` substring as whatever will be returned by this hook will be used for WP_Query? i.e. `product_collection_block_wp_query_vars`
- Should we consider using `args` instead of `vars` in the hook name? i.e. `product_collection_block_query_args`
- Should we consider adding `frontend` substring to communicate that this hook is intended for the frontend? i.e. `product_collection_block_frontend_query_vars`

What do you think? I just wanted to leave some thoughts to make sure we consider all options 🙂

> Proposal B: Introduce the rest_product_collection_block_query_args Filter.
> 

I am curious if we need to really pass $request as 2nd argument. I was looking at Product Collection code and AFAIU We are only using params from WP_REST_Request to update $args array. Probably we can just pass all params as 2nd argument instead of $request. What do you think? Maybe I am missing something?

Regarding name of this hook:

- Should we consider adding `wc_` prefix to avoid any conflicts with other plugins/themes? i.e. `wc_rest_product_collection_block_query_args`
- Should we consider adding `editor` substring to communicate that this hook is intended for the editor? i.e. `rest_product_collection_block_editor_query_args`
- Should we consider adding `wp_query` substring as whatever will be returned by this hook will be used for WP_Query? i.e. `rest_product_collection_block_wp_query_args`
- Do we even need `rest` prefix? Does it even matter for 3PDs POV?

---

Also, I believe 3PDs will still need to check if it's their collection or not. Right? Because this hook will be called for all collections on the page. Maybe we can also consider shipping a hook similar to `render_block_[block_name]`. It may look something like this: `product_collection_block_query_vars_<collection_name>` and `rest_product_collection_block_query_args_<collection_name>`. What do you think?

---

Also, maybe it would make sense to try to have both similar to each other but with a `frontend` and `editor` substring to indicate where they will be used. What do you think?

**Reply by xristos3490** [reply]
Hey @imanish003, and thanks for sharing your thoughts!

> But I just wanted to say that in 1st iteration, we can keep it as simple as possible and then iterate

Definitely! 

> Should we consider adding wp_query substring as whatever will be returned by this hook will be used for WP_Query? i.e. product_collection_block_wp_query_vars
Should we consider using args instead of vars in the hook name? i.e. product_collection_block_query_args
Should we consider adding frontend substring to communicate that this hook is intended for the frontend? i.e. product_collection_block_frontend_query_vars

My thinking is to keep things as simple as possible and maintain some familiarity from the Query Loop block. This means following a format of `[block_name]_block_query_args`, and brings me to the other suggestion:

> Should we consider adding wc_ prefix to avoid any conflicts with other plugins/themes? i.e. wc_rest_product_collection_block_query_args

Yes It seems that `product_collection_block_query_args` doesn't quite fit the `[block_name]_block_query_args` format, since the block is "woocommerce/product-collection". I like the idea of adding the "woocommerce_" prefix. What do you think about using "**woocommerce_product_collection_block_query_args**" and "**woocommerce_rest_product_collection_block_query_args**"?

> Do we even need rest prefix? Does it even matter for 3PDs POV?

I would prefer to keep things separate because each request is quite different, even though they could theoretically be condensed into a single hook. It's also a good practice to include the "rest" token when a hook is handling that context.

> How do you feel about also passing location?
> I am not sure if we need to pass $page as 3rd param. I believe we should set paged to $page in $query itself. We are doing it [here](https://github.com/woocommerce/woocommerce/blob/4253f2c10a985cdde05159eb27fbbe464a01f7d4/plugins/woocommerce/src/Blocks/BlockTypes/ProductCollection.php#L638) in Product Collection block. Does it make sense?

I have been considering this as well. While it seems like a great idea, it is still an experimental feature that is easily retrieved, right? I think the best course of action for now is to omit the $page argument and leave some room for further consideration in the future? After all, as you mentioned, the page information can be retrieved from the "paged" query argument.

What do you think?

**Reply by imanish003** [reply]
> My thinking is to keep things as simple as possible and maintain some familiarity from the Query Loop block. This means following a format of `[block_name]_block_query_args`, and brings me to the other suggestion:

I appreciate the reasoning behind maintaining consistency with the Query Loop block format. However, I have a slight concern:

While simplicity and familiarity are valuable, we should consider the potential for future divergence between the Product Collection block and the Query Loop block. If we adopt an identical naming convention now, it could lead to confusion for 3PDs down the line.

While I don't have a strong opinion, I think we should consider a more distinctive name. A clear, future-proof naming convention could prevent confusion for third-party developers if these blocks diverge later. 

> What do you think about using "woocommerce_product_collection_block_query_args" and "woocommerce_rest_product_collection_block_query_args"?

Agreed, those names work well. However, I think we should consider adding a clear distinction between frontend and editor hooks for better clarity. But again, I don't have strong opinion here. 

> I would prefer to keep things separate because each request is quite different, even though they could theoretically be condensed into a single hook. It's also a good practice to include the "rest" token when a hook is handling that context.

Makes sense to me.

> I have been considering this as well. While it seems like a great idea, it is still an experimental feature that is easily retrieved, right? I think the best course of action for now is to omit the $page argument and leave some room for further consideration in the future? After all, as you mentioned, the page information can be retrieved from the "paged" query argument.

Yeah! Sounds good to me. We can always iterate on this in the future 💯

---

Based on above thoughts, How do you feel about these:
- Frontend: `woocommerce_product_collection_frontend_query_args`
- Editor: `woocommerce_rest_product_collection_editor_query_args`

What do you think?

**Reply by xristos3490** [reply]
>Based on above thoughts, How do you feel about these:
Frontend: woocommerce_product_collection_frontend_query_args
Editor: woocommerce_rest_product_collection_editor_query_args
What do you think?

I like! Let's give this a few more days in case there are other opinions, and then I'll share the POC PR!

**Comment by ObliviousHarmony** [comment]
While working on the Related Products collection [I took the opportunity to explore what extensibility for the Product Collection block might look like](https://github.com/woocommerce/woocommerce/blob/1c055f4d282bd80af44a489945e6cbfdc7a76981/plugins/woocommerce/src/Blocks/BlockTypes/ProductCollection.php#L1826-L1871).

Rather than replacing the query loop hooks I decided to try for a more opinionated extensibility model. The way I see it, we're delivering the Product Collection as a consistent way to provide comprehensive product listing functionality. It has a lot of options for filtering both internally and through the use of other blocks that modify the query. This is _really_ powerful and I think it's a better idea to lean into integrating custom collections with this as opposed to encouraging people to modify the query after the collection's builder has run.

### Proposal

We currently have two functions that are provided raw query input. We have one for the frontend which receives the block and all related context and then another for the editor which updates a REST API query with additional arguments. In both of these cases we perform some minor transofrmation and reading of the input arguments into a consistent format that we can feed into `get_final_query_args()`. This feels like a good pattern and is something we should provide filtering behavior for.

- `woocommerce_product_collection_frontend_args`: This filter gets called [here](https://github.com/woocommerce/woocommerce/blob/1c055f4d282bd80af44a489945e6cbfdc7a76981/plugins/woocommerce/src/Blocks/BlockTypes/ProductCollection.php#L800C1-L804C4) right now _but_ I think it would probably be better in `build_frontend_query`.
- `woocommerce_product_collection_editor_args`: This filter gets called [here](https://github.com/woocommerce/woocommerce/blob/1c055f4d282bd80af44a489945e6cbfdc7a76981/plugins/woocommerce/src/Blocks/BlockTypes/ProductCollection.php#L655-L659). Note that it takes place before the preview so that it can provide the collection arguments to the preview.

The idea here is that we can provide a way for developers to drop custom arguments into the query builder. `$collection_args` is returned by any filters and then gets passed into the query builder. This lets developers perform any custom argument processing and then get that into the query builder agnostic of the frontend or editor context.

From there we provide two more filters:

- `woocommerce_product_collection_build_query`: This gets called [here](https://github.com/woocommerce/woocommerce/blob/1c055f4d282bd80af44a489945e6cbfdc7a76981/plugins/woocommerce/src/Blocks/BlockTypes/ProductCollection.php#L856-L867) and is passed the `$collection_args`. This is where developers can implement custom query generation. The great thing about doing this here is that the result gets passed into our own query merging functionality. This lets them use existing filtering alongside theirs without needing to do any additional parsing.
- `woocommerce_product_collection_preview`: This gets called [here](https://github.com/woocommerce/woocommerce/blob/1c055f4d282bd80af44a489945e6cbfdc7a76981/plugins/woocommerce/src/Blocks/BlockTypes/ProductCollection.php#L893-L897) and allows collections to define any custom preview behavior.

While it's true that this isn't as flexible as the above proposal, I think it's worth keeping in mind that collections are _variations_. They should follow the behavior of our block as closely as possible rather than trying to modify the query after we're totally done with it. I don't think there's a need for something lower-level either. Unless you pass query arguments, the builder doesn't actually _do_ anything.

[Here is a look at how all of this comes together](https://github.com/woocommerce/woocommerce/blob/1c055f4d282bd80af44a489945e6cbfdc7a76981/plugins/woocommerce/src/Blocks/BlockTypes/ProductCollection.php#L1826-L1871). Substitute the functions for hooks in your mind 😄

**Reply by kmanijak** [reply]
I like the opinionated approach more. While mimicking Query Loop and exposing low-level filters is one way of doing it, we don't have to follow the same approach.

BTW, I gave it a try with [Upsells collection](https://github.com/woocommerce/woocommerce/pull/51374) and it worked out nicely. The interface seems clear and **gives less responsibility to the developer** which I think is a key! In addition, we could even provide some default `$frontend_args` and `$editor_args` functions for common cases (like in out case - Related Products, Upsells, Cross-sells are going to have the same logic for these).

**Reply by imanish003** [reply]
> Rather than replacing the query loop hooks I decided to try for a more opinionated extensibility model.

I like the idea of providing developers with a more opinionated and easier-to-use interface. It's a good approach to reducing complexity and providing a clear way for developers to extend the functionality of the Product Collection block 💪🏻

**Question:** For now we have extensibility points i.e. `query_loop_block_query_vars` and `rest_product_query`, which are available for 3PDs to use. If we go with the proposed solution, will it replace the existing hooks or will it co-exist?

AFAIU we won't be removing existing hooks i.e. `query_loop_block_query_vars` and `rest_product_query`. But I believe we won't be promoting them in our documentation. Those exist for some reasons i.e. 
- `query_loop_block_query_vars` because we are using core blocks like Pagination.
- `rest_product_query` because we use REST API to fetch products in the editor.

Therefore, we will still need to keep them while adding new hooks. It's important to note that these existing hooks are not specific to the Product Collection block. The new hooks we're proposing would be more tailored to the Product Collection block's specific needs.

---

Also, IMO, any new extensibility points should be prefixed with `__experimental_` to make it clear that these hooks are experimental and may change in the future. This way, we will have a bit more freedom to make these hooks more powerful and flexible in the future without worrying about breaking changes.

---

> The idea here is that we can provide a way for developers to drop custom arguments into the query builder. `$collection_args` is returned by any filters and then gets passed into the query builder. This lets developers perform any custom argument processing and then get that into the query builder agnostic of the frontend or editor context.

I like the idea 🚀

---

In conclusion, I like this idea of opinionated hooks. However, I am not sure that the signature of all these new hooks will be the same as in the "Related Products" PR. But I believe this can be discussed during the PR review, which will add these new extensibility points. Also, As I suggested, we must add `__experimental_` prefix to all these new hooks to clarify that these hooks are experimental and may change in the future. IMO We should wait for some time before implementing these new hooks, probably till all core collections are in place that uses [register_collection_handlers](https://github.com/woocommerce/woocommerce/blob/add/49335-related-products-collection/plugins/woocommerce/src/Blocks/BlockTypes/ProductCollection.php#L1815). This way, we will be able to test these new extensibility points with our core collections. What do you think?

**Reply by xristos3490** [reply]
I'd like to share a few thoughts on two different topics:

### 1. Minimizing the area of impact of these filters

I generally support this idea. I was unsure whether we should handle the query arguments as standalone arguments or pass the entire array after the core has built it. I don't have a strong preference, but it might be a good idea to start with a narrow impact. Later on, if more people are in favor of this, we can provide another hook as a replacement for the `query_loop_block_query_args` hook. We can always introduce a e.g., `woocommerce_product_collection_frontend_query_args_all` hook later to compensate.

However, for the sake of argument, what should someone do **if they need to add another inspector setting to every PC block**? How can they handle the query without a hook for all arguments?

That brings me to the second topic:

### 2. Keeping it simple

I was looking at the `register_collection_handlers()` helper, and I'm worried it's biased towards our specific needs. Let me explain.

What we should focus on here are three potential areas of extensibility based on the block context: (a) the frontend, (b) the editor's side, and (c) optionally, the preview mode if it's enabled within the editor. In each of these scenarios, our goal is simple: "Get some query arguments, and return a modified set of them." If we make it more complex than that, it could effectively increase the barrier of entry here.

Therefore, a third-party developer only needs to understand that they have **to modify query arguments**. If we add another layer that gets passed to another layer (referring to get_collection_args + get_query_args), it defeats the purpose of simplicity, doesn't it? I believe that if a third-party developer realizes that the query building in the frontend and editor is the same, it's quite easy to create a reusable function in PHP. Why do we need to make that architectural decision for them?

**Final Thoughts**

I think both (1) and (2) points can be discussed _separately_, as they bring two very different improvements/standards in this discussion. Moreover, for the purpose of streamlining the hooks for custom collections, we could also consider either: 
- Adding "collection-specific" hooks (as mentioned by @imanish003 ), or; 
- Passing the collection name as an argument in these hooks by default.

What do you all think?

**Reply by imanish003** [reply]
> However, for the sake of argument, what should someone do if they need to add another inspector setting to every PC block? How can they handle the query without a hook for all arguments?
> 

If we go with Christopher's proposal

- `editor.BlockEdit` hook will be used to add new inspector controls UI.
    - If inspector control requires an attribute, it can be added to the "query" attribute.
- `editor_args` and `frontend_args` callback functions will be used to pass new attributes to the `build_query` callback function.
- `build_query` callback function will use the attribute to modify the query.

Does it make sense? Please let me know in case I am missing something.

> I was looking at the register_collection_handlers() helper, and I'm worried it's biased towards our specific needs
> 

Yeah, but I think the goal is to test these with our core collection. Because, IMO, core collections and 3P collections should work the same way. If a solution works for a core collection, I believe there is a high chance that it will work for 3P collections as well. Isn't it? 🤔

> Why do we need to make that architectural decision for them?
> 

That's a great question! You raise an important point about architectural decisions and how they impact extensibility for third-party developers. I think it's worth carefully considering the tradeoffs between providing helpful abstractions versus keeping things simple and flexible.

After reading your comment, I am inclining more towards solution which you proposed [here](https://github.com/woocommerce/woocommerce/pull/51076#discussion_r1764692718) i.e.:

```
$this->register_collection_handlers(
    build_frontend_query,
    build_editor_query,
    build_preview_query (optional)
);

```

This will allow us to build on top of current hooks. As we discussed in above [comment](https://github.com/woocommerce/woocommerce/discussions/51018#discussioncomment-10598961):

- Frontend: `woocommerce_product_collection_frontend_query_args`
- Editor: `woocommerce_rest_product_collection_editor_query_args`

These hooks can integrate quite well with `build_frontend_query` and `build_editor_query` handlers. This can actually work pretty well if we can pass additional arguments to these hooks like:

- Collection name
- Location
- In case, collection is using `usesReference`, we can pass additional argument with  `productID`, `productIds`. This way 3PDs won't need to manually calculate it. Both "Related Products" and "Upsells" collection are using `productId` or `productIds`, so it's useful for core collections too.

**Reply by ObliviousHarmony** [reply]
Thanks @xristos3490 for the thoughtful reply!

> I don't have a strong preference, but it might be a good idea to start with a narrow impact.

This is exactly the motivation. We can _always_ add broad filters.

> what should someone do if they need to add another inspector setting to every PC block?

Right now our client-side registration API heavily leans into the idea that each collection is its own thing. It has query presets, however, there's no mechanism in place to include a control on every block. I don't think this point is directly related to server-side filtering so much as it is a limitation of the current client-side API.

> How can they handle the query without a hook for all arguments?

Could you expand on what you mean here?

> I was looking at the register_collection_handlers() helper, and I'm worried it's biased towards our specific needs.

To be fair, this is mostly the point. Our query builder implementation makes the (reasonable) assumption that the frontend and editor have different request formats. This content is then fed into a builder that is agnostic to the request source. The exception there is preview mode but that's something else to handle entirely. My rationale is that _most_ collections will end up re-implementing that same mechanism either verbatim or by having a single function they bind to both hooks and handle it circumstantially. (cc @imanish003)

> Therefore, a third-party developer only needs to understand that they have to modify query arguments

There's nothing wrong with _only_ using the `woocommerce_product_collection_build_query` filter. The other two filters only exist to allow for the same request input processing that we utilize ourselves. A collection can add custom `query` arguments and then pass them in, and if there's some translation that needs done, we can take care of it there.

> I believe that if a third-party developer realizes that the query building in the frontend and editor is the same, it's quite easy to create a reusable function in PHP. Why do we need to make that architectural decision for them?

I believe that it's our role to provide an extensibility model that encourages good practices, even if it requires a _little_ bit more code.

> Passing the collection name as an argument in these hooks by default

Note that the `$collection_args` is pre-populat

ed with the collection name. This means that all of the filters are passed the name and can use it to decide whether or not it needs to be applied to that collection. This is neat because it also supports applying the same filter to multiple collections (or all!).

**Comment by helgatheviking** [comment]
+1 I am so here for custom query filters instead of using `query_loop_block_query_vars` and needing to conditionally test if you are a Product Collection block.
